lib/fluent/plugin/out_google_cloud.rb

```diff
@@ -665,14 +678,140 @@ def write(chunk)
           rescue Google::Apis::ClientError => error
             # Most ClientErrors indicate a problem with the request itself and
             # should not be retried.
-            increment_dropped_entries_count(entries_count)
-            @log.warn "Dropping #{entries_count} log message(s)",
-                      error_class: error.class.to_s, error: error.to_s
+            error_details_map = {}
+            if @partial_success
+              error_details_map = construct_error_details_map(
+                extract_log_entry_errors(error))
+            end
+            if error_details_map.empty?
+              increment_dropped_entries_count(entries_count)
+              @log.warn "Dropping #{entries_count} log message(s)",
+                        error_class: error.class.to_s, error: error.to_s
+            else
+              error_details_map.each do |(error_code, error_message), indexes|
+                partial_error_count = indexes.length
+                increment_dropped_entries_count(partial_error_count)
+                @log.warn "Dropping #{partial_error_count} log message(s)",
+                          error_code: "google.rpc.Code.#{error_code}",
+                          error: error_message
+              end
+            end
           end
         end
       end
     end
 
+    # Extract a map of error details from an potentially partially successful
+    # request. The keys in this map are the indexes of the failed log entries in
+    # the original json payload. The values in this map are the error code and
+    # error message.
+    #
+    # A sample error.body looks like:
+    # {
+    #   "error": {
+    #     "code": 403,
```

Comment:
Is this really 403? I thought that translated to `Google::Apis::AuthorizationError`, which would have been caught earlier...

Reply:
`401` maps to [Google::Apis::AuthorizationError](http://www.rubydoc.info/github/google/google-api-ruby-client/Google/Apis/AuthorizationError). Every other `4XX` maps to [Google::Apis::ClientError](http://www.rubydoc.info/github/google/google-api-ruby-client/Google/Apis/ClientError).

This sample is the output from this piece of code:

```ruby
$ cat non_grpc.rb
require 'google/apis/logging_v2'
require 'googleauth'

client = Google::Apis::LoggingV2::LoggingService.new
client.authorization = Google::Auth.get_application_default(
  'https://www.googleapis.com/auth/logging.write')

resource = Google::Apis::LoggingV2::MonitoredResource.new(
  type: 'gce_instance',
  labels: {}
)

entries = [
  Google::Apis::LoggingV2::LogEntry.new(
    {
      resource: resource,
      log_name: 'projects/gapic-test/logs/my-log',
      text_payload: 'pass'
    }
  ),
  Google::Apis::LoggingV2::LogEntry.new(
    {
      resource: resource,
      log_name: 'projects/gapic-test/logs/bad:log',
      text_payload: 'fail'
    }
  )
]

write_request = \
  Google::Apis::LoggingV2::WriteLogEntriesRequest.new(
    log_name: 'projects/gapic-test/logs/my-log',
    resource: resource,
    labels: {},
    partial_success: true,
    entries: entries)

begin
  client.write_entry_log_entries(
    write_request,
    options: {
      api_format_version: '2'
    }
  )
rescue Google::Apis::ClientError => error
  puts "error:\n#{error.inspect}\n"
  puts "error body =\n#{error.body}\n"
end
```

```
$ ruby non_grpc.rb
error:
#<Google::Apis::ClientError: PERMISSION_DENIED: User not authorized.>
error body =
{
  "error": {
    "code": 403,
    "message": "User not authorized.",
    "status": "PERMISSION_DENIED",
    "details": [
      {
        "@type": "type.googleapis.com/google.logging.v2.WriteLogEntriesPartialErrors",
        "logEntryErrors": {
          "0": {
            "code": 7,
            "message": "User not authorized."
          },
          "1": {
            "code": 3,
            "message": "Log name contains illegal character :"
          }
        }
      },
      {
        "@type": "type.googleapis.com/google.rpc.DebugInfo",
        "detail": "[ORIGINAL ERROR] generic::permission_denied: User not authorized. [google.rpc.error_details_ext] { message: \"User not authorized.\" details { type_url: \"type.googleapis.com/google.logging.v2.WriteLogEntriesPartialErrors\" value: \"\\n\\034\\010\\000\\022\\030\\010\\007\\022\\024User not authorized.\\n-\\010\\001\\022)\\010\\003\\022%Log name contains illegal character :\" } }"
      }
    ]
  }
}
```

Reply:
Will we always get a 403 with partial errors, or is this just an example?

Reply:
This is just an example. I just added some comment to clarify:

```
    # Note that the root level "code", "message" and "status" simply matches the
    # root cause of the first failed log entry. For example, if we switch the
    # order of the log entries, then we will get:
    # {
    #    "error" : {
    #       "code" : 400,
    #       "message" : "Log name contains illegal character :",
    #       "status" : "INVALID_ARGUMENT",
    #       "details": ......
    #    }
    # }
    # We will ignore it anyway and look at the details instead which includes
    # info for all failed log entries.
```

Reply:
I just confirmed that it's not possible to get a 401 on only one of the entries -- it always applies to the whole request. So we don't have to worry about parsing partial errors when catching an `AuthorizationError`. But what about `ServerError`? What if the first entry has a 500?

Reply:
From the documentation of `partial_success` in the `Stackdriver Logging API` site:

>partial success: Optional. Whether valid entries should be written even if some other entries fail due to INVALID_ARGUMENT or PERMISSION_DENIED errors. If any entry is not written, then the response status is the error associated with one of the failed entries and the response includes error details keyed by the entries' zero-based index in the entries.write method.

It sounds like only `INVALID_ARGUMENT` or `PERMISSION_DENIED` will trigger partial success. I took a look at the extra implementation and unit tests of the Logging API write request handler, it also suggests so.